benchmark/utils/general.hpp

```diff
@@ -434,4 +460,130 @@ gko::remove_complex<ValueType> compute_max_relative_norm2(
+ * ```
+ * auto timer = get_timer(...);
+ * IterationControl ic(timer);
+ * for(auto stat: ic.[warmup_run|run]()){
+ *   timer->start();  // has to be the same timer as passed to ic
+ *   // execute benchmark
+ *   timer->stop();
+ * }
+ * ```
+ * At the beginning of both methods, the timer is reset.
+ *
+ */
+class IterationControl {
+    using IndexType = unsigned int;  //!< to be compatible with GFLAGS type
+
+    class run_control;
+
+public:
+    // criteria to stop adaptive benchmark run
+    struct status {
+        std::shared_ptr<Timer> timer = nullptr;
+
+        IndexType min_it = 0;
+        IndexType max_it = 0;
+        double max_runtime = 0.;
+
+        IndexType cur_it = 0;
+
+        // returns true (i.e. run is complete) if:
+        // - the minimum number of iteration is reached
+        // - and either:
+        //   - the maximum number of repetitions is reached
+        //   - the total runtime is above the threshold
+        bool is_finished()
+        {
+            return cur_it >= min_it &&
+                   (cur_it >= max_it || timer->get_total_time() >= max_runtime);
+        }
+        bool is_last_iteration()
+        {
+            return status{timer, min_it, max_it, max_runtime, cur_it + 1}
+                .is_finished();
+        }
+    };
+
+    explicit IterationControl(std::shared_ptr<Timer> timer) : timer_(timer)
+    {
+        status_warmup_ = {timer, FLAGS_warmup, FLAGS_warmup, 0., 0};
+        if (FLAGS_repetitions == "auto") {
+            status_run_ = {timer, FLAGS_min_repetitions, FLAGS_max_repetitions,
+                           FLAGS_min_runtime};
+        } else {
+            const auto reps =
+                static_cast<unsigned int>(std::stoi(FLAGS_repetitions));
+            status_run_ = {timer, reps, reps, 0., 0};
+        }
+    }
+
+    IterationControl() = default;
+    IterationControl(const IterationControl &) = default;
+    IterationControl(IterationControl &&) = default;
+
+    run_control warmup_run()
+    {
+        timer_->clear();
+        status_warmup_.cur_it = 0;
+        return run_control{status_warmup_};
+    }
+
+    run_control run()
+    {
+        timer_->clear();
+        status_run_.cur_it = 0;
+        return run_control{status_run_};
+    }
+
+private:
+    // class to be used in range-based for
+    struct run_control {
+        struct iterator {
+            iterator operator++()
+            {
+                cur_info.cur_it++;
+                return *this;
+            }
+
+            status operator*() const { return cur_info; }
+
+            bool operator!=(const iterator &)
+            {
+                return !cur_info.is_finished();
+            }
+
+            status cur_info;
+        };
+
+        iterator begin() const { return iterator{info}; }
+
+        // not used, could potentially used in c++17 (with different type)
+        iterator end() const { return iterator{}; }
```

Comment:
I guess you are talking about [iterators sentinels](https://www.foonathan.net/2020/03/iterator-sentinel/), which are too beautiful a concept not to mention here :)